### ISSUE TITLE
Provide a modern wrapper for C++

### ISSUE THREAD

**Issue by ysc3839** [MicrosoftEdge/WebView2Feedback#316]
Current there's only a classic COM API for C++. It's bad for writing simple and safe code.

I hope there will be a modern C++ wrapper, like the `Microsoft.Web.WebView2.Core` .Net wrapper.
So that we can take advantages of async function, smart pointers, exception based error handling and so on.

There's already some user asking for such feature. For example #99, https://github.com/microsoft/cppwinrt/issues/562 and https://github.com/MicrosoftEdge/WebView2Samples/issues/30.

Personally I think it's good to use C++/WinRT to do this job, using a Runtime Component.

**Comment by oggy22** [comment]
Thank you for your feedback. I also think it's important to have modern C++ APIs. Can you provide some examples where you think  we can improve our APIs to be more modern C++?

**Comment by ysc3839** [comment]
@oggy22 Just like C++/WinRT and .Net. For example:
```
#include "WebView2.h"
using namespace Microsoft.Web.WebView2.Core;

auto env = co_await CoreWebView2Environment::CreateAsync(nullptr, nullptr, nullptr);
auto webviewController = co_await env.CreateCoreWebView2ControllerAsync(hWnd);
auto webviewWindow = webviewController.CoreWebView2();

auto Settings = webviewWindow.Settings();
Settings.IsScriptEnabled(true);
Settings.AreDefaultScriptDialogsEnabled(true);
Settings.IsWebMessageEnabled(true);

RECT bounds;
GetClientRect(hWnd, &bounds);
webviewController.Bounds(bounds);

webviewWindow.Navigate(L"https://www.bing.com/");
```

**Comment by liminzhu** [comment]
@ysc3839 good feedback. We're actually adding a WinRT wrapper to our COM APIs that you would be able to use it through C++/WinRT.

**Comment by john-preston** [comment]
@liminzhu Hi. What is the status of WinRT wrappers for WebView2?

It's a bit strange to use new modern WinRT wrappers for the old WebViewControl based on the old Edge browser, but to be forced to use old plain COM code with some helpers for the new WebView2 based on the new Edge browser.

**Comment by liminzhu** [comment]
@john-preston sorry for the late reply. If you're using WinUI3 flavor of WebView2, you have access to the WinRT wrapper. We haven't really explored WinRT wrapper outside of the context of WinUI.

**Comment by champnic** [comment]
The WinRT wrapper is available in our nuget package under "runtimes\win-<arch>\native_uap\Microsoft.Web.WebView2.Core.dll", with a winmd in the "lib" folder.

**Comment by john-preston** [comment]
@liminzhu I'm using WebView2 in a native C++ app that doesn't use WinUI3, just uses the WebView2 NuGet package.

@champnic Hi, I've tried generating C++/WinRT headers from the Microsoft.Web.WebView2.Core.winmd file and I encountered some problems. Looks like I cannot replace using old WRL helpers with the C++/WinRT headers.

For example, if I use "CoreWebView2CustomSchemeRegistration" implementation from the "build/native/include/WebView2EnvironmentOptions.h", which is based on WRL, it can be constructed from a scheme name and has a "SetAllowedOrigins" method. But in the generated headers there is nothing related to AllowedOrigins at all.

Or do I understand something wrong and there is no way to register custom scheme using the C++/WinRT headers for WebView2 interaction, only by implementing the ICoreWebView2CustomSchemeRegistration ourself?
